sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/MapperSpec.kt

```diff
@@ -19,158 +19,200 @@ import com.squareup.javapoet.ArrayTypeName
 
-  private fun mapperMethod(): MethodSpec {
-    val mapReturn = CodeBlock.builder().add("$[return $CREATOR_FIELD.create(\n")
-
-    for (column in table.column_def()) {
-      if (column != table.column_def(0)) mapReturn.add(",\n")
-      if (column.isHandledType) {
-        mapReturn.add(cursorMapper(column))
-      } else {
-        if (column.isNullable) {
-          mapReturn.add("$CURSOR_PARAM.isNull(" +
-              "$CURSOR_PARAM.getColumnIndex(${column.constantName(nameAllocator)})) ? null : ")
-        }
-        mapReturn.add("${column.adapterField(nameAllocator)}.$MAP_FUNCTION(" +
-            "$CURSOR_PARAM, $CURSOR_PARAM.getColumnIndex(${column.constantName(nameAllocator)}))")
-      }
+    sortedResultsMap(
+        { columnName, indexedValue -> indexedValue.cursorGetter() },
+        { tableName, table -> table.cursorGetter() }
+    ).forEachIndexed { i, codeBlock ->
+      if (i != 0) mapReturn.add(",\n")
+      mapReturn.add(codeBlock)
     }
 
-    return MethodSpec.methodBuilder(MAP_FUNCTION)
+    val mapMethod = MethodSpec.methodBuilder("map")
         .addAnnotation(Override::class.java)
-        .addAnnotation(NONNULL_TYPE)
-        .addModifiers(PUBLIC)
+        .addAnnotation(MapperSpec.NONNULL_TYPE)
+        .addModifiers(Modifier.PUBLIC)
         .returns(TypeVariableName.get("T"))
         .addParameter(ParameterSpec.builder(CURSOR_TYPE, CURSOR_PARAM)
             .addAnnotation(NONNULL_TYPE)
             .build())
         .addCode(mapReturn.add("$]\n);\n").build())
-        .build()
-  }
 
-  private fun cursorMapper(
-      column: SqliteParser.Column_defContext,
-      columnName: String = column.constantName(nameAllocator)
-  ): CodeBlock {
-    val code = CodeBlock.builder()
-    if (column.isNullable) {
-      code.add("$CURSOR_PARAM.isNull($CURSOR_PARAM.getColumnIndex($columnName)) ? null : ")
-    }
-    return code.add(column.cursorGetter("$CURSOR_PARAM.getColumnIndex($columnName)")).build()
+    return mapper.addMethod(constructor.build())
+        .addMethod(mapMethod.build())
   }
 
-  private fun creatorInterface(): TypeSpec {
-    val create = MethodSpec.methodBuilder(CREATOR_METHOD_NAME)
-        .returns(TypeVariableName.get("R"))
-        .addModifiers(PUBLIC, ABSTRACT)
-
-    for (column in table.column_def()) {
-      create.addParameter(column.javaType, column.methodName(nameAllocator))
-    }
-
-    return TypeSpec.interfaceBuilder(creatorClassName.simpleName())
-        .addTypeVariable(TypeVariableName.get("R", table.interfaceClassName))
-        .addModifiers(PUBLIC)
-        .addMethod(create.build())
-        .build()
+  private fun addFactoryField(
+      mapper: TypeSpec.Builder,
+      constructor: MethodSpec.Builder,
+      typeVariable: String,
+      interfaceClass: ClassName
+  ): String {
+    val factoryFieldname = "${interfaceClass.simpleName().decapitalize()}$FACTORY_NAME"
+    mapper
+        .addField(ParameterizedTypeName.get(interfaceClass.nestedClass(FACTORY_NAME),
```

Comment:
cache this TypeName creation in a local. you're doing it twice, and it clutters these lines.